keps/3258-delayed-admission-check-retries/README.md

```diff
@@ -0,0 +1,545 @@
+
+The admission check controller can:
+
+1. Detect control plane stress (high API server latency, etcd performance issues)
+2. Set workloads to `Retry` with delays proportional to the stress level
+3. Prevent thundering herd effects by spreading retry times
+
+#### Story 3: Multi-tenancy with Different Retry Strategies
+
+As a platform engineer supporting multiple teams, I want different admission checks to have independent retry strategies based on their specific requirements without interfering with each other.
+
+For example:
+
+- Security scanning admission check: retry after 5 minutes if scanner is busy
+- Resource provisioning check: retry after 1 minute if capacity is being added
+- Compliance validation check: retry after 10 minutes if external audit service is down
+
+### Notes/Constraints/Caveats
+
+- The retry delay is advisory; the actual delay might be slightly longer due to reconciliation timing
+- Workloads with very long retry delays might appear to be "stuck" to users
+- When multiple admission checks specify different retry delays, the workload will not be retried until all delays have expired
+- The retry mechanism does not guarantee strict FIFO ordering when delays are involved
+- There is a potential race condition where Kueue might reset admission check states before all controllers have set their retry delays
+
+### Risks and Mitigations
+
+**Risk**: Breaking backward compatibility with existing admission check controllers\
+**Mitigation**: The new fields are optional; controllers that don't set them will maintain current behavior
+
+**Risk**: Workloads appearing stuck due to long retry delays \
+**Mitigation**: Add clear status messages that indicate the retry time, expose metrics to monitor retry delays, and document the recommended maximum delay values.
+
+**Risk**: Resource leakage if workloads are forgotten during long delays \
+**Mitigation**: Ensure proper cleanup mechanisms and consider maximum delay limits
+
+**Risk**: Race conditions when multiple admission checks set delays \
+**Mitigation**: Allow and support admission checks to update the state after a workload has been evicted.
+
+## Design Details
+
+### API Changes
+
+Add new fields to `AdmissionCheckState` in `workload_types.go`:
+
+```go
+type AdmissionCheckState struct {
+    // Existing fields remain unchanged...
+
+    // RetryAfter indicates when this check should be retried.
+    // When State=Retry, admission check controllers should set this
+    // to implement delays between retry attempts.
+    //
+    // If nil when State=Retry, Kueue will retry immediately.
+    // If set, Kueue will not re-evaluate until this time has passed.
+    //
+    // Example: To retry in 5 minutes:
+    //   retryAfter = &metav1.Time{Time: time.Now().Add(5*time.Minute)}
+    //
+    // +optional
+    RetryAfter *metav1.Time `json:"retryAfter,omitempty"`
```

Comment:
This is not necessary if we take this more declarative approach: https://github.com/kubernetes-sigs/kueue/pull/6210#issuecomment-3155482372